test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,456 @@
+        elif type(python_arg) == torch.Tensor:
+            return CppArg(
+                type='torch::Tensor',
+                value='torch::empty({})'.format(str(list(python_arg.shape)).replace('[', '{').replace(']', '}')))
+        else:
+            raise RuntimeError(
+                "{} is not a supported arg type for C++ module methods".format(type(python_default_value)))
+
+    def _compile_cpp_code_inline(self, name, cpp_sources, functions):
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=name,
+            cpp_sources=cpp_sources,
+            functions=functions,
+            verbose=False,
+        )
+        return cpp_module
+
+    # This tests that Python and C++ torch.nn modules have matching constructor arg names and types.
+    def _test_torch_nn_module_ctor_args(self, module_name):
+        python_module_class = getattr(torch.nn, module_name)
+        module_metadata = torch_nn_modules.module_metadata_map[module_name]
+        cpp_default_constructor_args = module_metadata['cpp_default_constructor_args']
+
+        cpp_module_option = 'torch::nn::{}Options{}'.format(module_name, cpp_default_constructor_args)
+        if PY2:
+            init_arg_spec = inspect.getargspec(python_module_class.__init__)
+        else:
+            init_arg_spec = inspect.getfullargspec(python_module_class.__init__)
+        init_kwargs_defaults = init_arg_spec.defaults
+        init_kwargs = init_arg_spec.args[-len(init_kwargs_defaults):]
+        for arg_name, python_default_value in zip(init_kwargs, init_kwargs_defaults):
+            cpp_module_option += '.{}({})'.format(arg_name, self._python_arg_to_cpp_arg(python_default_value).value)
+
+        cpp_sources = TORCH_NN_MODULE_COMMON_TEST_HARNESS + module_metadata.get('cpp_sources', '')
+        cpp_sources += TORCH_NN_MODULE_TEST_CTOR_ARGS.substitute(
+            module_name=module_name,
+            module_qualified_name='torch::nn::{}'.format(module_name),
+            module_option=cpp_module_option)
+        cpp_test_name = module_name + '_test_ctor_args'
+        cpp_module = self._compile_cpp_code_inline(
+            name=cpp_test_name, cpp_sources=cpp_sources, functions=cpp_test_name)
+
+        getattr(cpp_module, cpp_test_name)()
+
+    def _test_torch_nn_module_variant(self, test_params):
+        torch_nn_test_methods = ['init', 'forward', 'backward']
+
+        def generate_test_cpp_sources(test_params, template, extra_stmts):
+            example_inputs = test_params.example_inputs
+            input_arg_types = [self._python_arg_to_cpp_arg(arg).type for arg in example_inputs]
+            input_arg_declarations = ',\n'.join(
+                ['{} arg{}'.format(input_arg_types[i], str(i)) for i in range(len(input_arg_types))])
+            input_args = ',\n'.join(['arg{}'.format(str(i)) for i in range(len(input_arg_types))])
```

Comment:
It might be nicer to create `input_args` first, and then do 
```
input_arg_declarations = ['{} {}.format(typ, name) for typ, name in zip(input_arg_types, input_args)]
```
to avoid the duplications